### ISSUE TITLE
Bounds and constraints of type variables cannot be generic

### ISSUE THREAD

**Issue by mtshiba** [astral-sh/ty#1839]
### Summary

In the current Python type specification, bounds and constraints for type variables must be concrete types and cannot contain type variables (even though it's possible at runtime).

[peps.python.org/pep-0695#type-parameter-scopes](https://peps.python.org/pep-0695/#type-parameter-scopes)

So, all of the following cases should be reported as errors:

```py
from typing import TypeVar

# OK
T1 = TypeVar("T1", bound=list[int])
T2 = TypeVar("T2", list[int], str)

U = TypeVar("U")
# error
T3 = TypeVar("T3", bound=U)
# error
T4 = TypeVar("T4", bound=list[U])
# error
T5 = TypeVar("T5", bound=list["T5"])
# error
T6 = TypeVar("T6", U, str)
# error
T7 = TypeVar("T7", list["T7"], str)
```

```py
# OK
def _[T: list[int]](): ...
def _[T: (list[int], str)](): ...

# error
def _[U, T: U](): ...
# error
def _[U, T: list[U]](): ...
# error
def _[T: list[T]](): ...
# error
def _[U, T: (U, str)](): ...
# error
def _[T: (list[T], str)](): ...
```

However, this is difficult to implement straightforwardly in the current implementation of ty, because bounds/constraints on type variables are lazily evaluated.

**Comment by AlexWaygood** [comment]
> However, this is difficult to implement straightforwardly in the current implementation of ty, because bounds/constraints on type variables are lazily evaluated.

Evaluation of class bases is also deferred, so most checks that would involve iterating over a class's bases or MRO are done after the rest of the scope has been inferred. We can probably do something similar for checking type variables? See https://github.com/astral-sh/ruff/blob/ff7086d9ad24729891754ba597ca5abc513e4858/crates/ty_python_semantic/src/types/infer/builder.rs#L581

**Comment by mtshiba** [comment]
> Evaluation of class bases is also deferred, so most checks that would involve iterating over a class's bases or MRO are done after the rest of the scope has been inferred. We can probably do something similar for checking type variables? See [astral-sh/ruff@`ff7086d`/crates/ty_python_semantic/src/types/infer/builder.rs#L581](https://github.com/astral-sh/ruff/blob/ff7086d9ad24729891754ba597ca5abc513e4858/crates/ty_python_semantic/src/types/infer/builder.rs#L581)

Ah, maybe we can use this pattern. I'll give it a try.

**Comment by AlexWaygood** [comment]
Fixed in https://github.com/astral-sh/ruff/pull/22949 and https://github.com/astral-sh/ruff/pull/22982
